### PR TITLE
[mlir][memref] Remove incorrect `memref.transpose` fold

### PR DESCRIPTION
**Pull request by MacDue**
This folded casts into `memref.transpose` without updating the result type of the transpose op, which resulted in IR that failed to verify for statically sized memrefs.

i.e.

```mlir
%cast = memref.cast %0 : memref<?x4xf32> to memref<?x?xf32>
%transpose = memref.transpose %cast : memref<?x?xf32> to memref<?x?xf32>
```

would fold to:

```mlir
// Fails verification:
%transpose = memref.transpose %cast : memref<?x4xf32> to memref<?x?xf32>
```

### CODE REVIEW

File: mlir/test/Dialect/MemRef/canonicalize.mlir [hunk 1]
```diff
@@ -1023,3 +1023,16 @@ func.func @fold_identity_transpose(%arg0: memref<1x2x3x4x5xf32>) -> memref<1x2x3
   // CHECK: return %[[arg0]]
   return %1 : memref<1x2x3x4x5xf32>
 }
+
+// -----
+
+// CHECK-LABEL: func @cannot_fold_transpose_cast(
+//  CHECK-SAME:     %[[arg0:.*]]: memref<?x4xf32, strided<[?, ?], offset: ?>>
+func.func @cannot_fold_transpose_cast(%arg0: memref<?x4xf32, strided<[?, ?], offset: ?>>) -> memref<?x?xf32, strided<[?, ?], offset: ?>> {
```
Comment (c-rhodes) [inline mlir/test/Dialect/MemRef/canonicalize.mlir:1031]: nit: is the strided/offset metadata necessary to reproduce or could it be removed to simplify this test a little?

**Review by c-rhodes — APPROVED** [review]
thanks for fix Ben, makes sense to me, just left one minor nit/question, otherwise LGTM, cheers

**Review by ubfx — APPROVED** [review]
I agree it makes sense to remove this fold. I just wonder what the original reason was for having it... Maybe it used to make sense before we had a verifier.

**Comment by MacDue** [comment]
>  I just wonder what the original reason was for having it... Maybe it used to make sense before we had a verifier.

I'm not sure... but given there's no tests covering it, I think it's safe to remove.

**Pull request merged**